operator/builtin/transformer/recombine/recombine.go

```diff
@@ -230,45 +254,66 @@ func (r *RecombineOperator) matchIndicatesLast() bool {
 
 // addToBatch adds the current entry to the current batch of entries that will be combined
 func (r *RecombineOperator) addToBatch(_ context.Context, e *entry.Entry) {
-	if len(r.batch) >= r.maxBatchSize {
+	var s string
+
+	if r.sourceIdentifier.FieldInterface != nil {
+		err := e.Read(r.sourceIdentifier, &s)
+		if err != nil {
+			r.Warn("entry does not contain the source_identifier, so it may be pooled with other sources")
+			s = "DefaultSourceIdentifier"
+		}
+	} else {
+		s = "DefaultSourceIdentifier"
+	}
+	if r.batchSize >= r.maxBatchSize {
```

Comment:
The way you've chosen to use this is interesting. We are batching entries by source, but counting them all together as an overall maximum. My first instinct was to suggest that we just allow each batch to hit a max size, but the overall max has a couple nice benefits for memory usage. For one, it roughly limits the overall size of memory that may be need to be allocated by the operator. The other benefit is that it provides a simple way to garbage collect old sources over time. 

All that said, I wonder if this could lead to difficult-to-explain behavior in some cases. For example, as a user, I think I would expect that `max_batch_size` will only place a per-source limit on the number of entries that would be combined into a single message. Let's say I know my system splits all logs into two parts, so I just set `max_batch_size: 2`. Now if I have more than one source, I will have lots of premature flushes.

An alternate design, which I think would solve all the same problems while being a little easier to understand:
- Let's apply `max_batch_size` to each source individually.
  - This allows us to remove `r.batchSize` and just depend on `len(batchMap[s])`
- Add another config setting called `max_sources` with a reasonable default, maybe 1000
  - At the end of `addToBatch`, check `if len(batchMap) > r.maxSources`
    - If so, we need to determine flushing logic. I think it's probably ok to just flush all at this point
- Whenever we flush entries for a source, `delete(batchMap, s)` to ensure we are tracking number of sources correctly